### PR TITLE
add CockroachDB SQL schema

### PR DESCRIPTION
**Pull request by Hades32**
Thise adds a copy of the Postgres v12 schema with minimal changes to make it compatible with CRDB v23.1+

Fixes https://github.com/temporalio/temporal/issues/4136

Relates to https://github.com/temporalio/temporal/issues/4180

### CODE REVIEW

**Comment by Hades32** [comment]
As this is a copy, the diff is hard to see. This is what it looks compared to pg v12
```patch
diff --git a/schema/postgresql/v12/visibility/versioned/v1.2/advanced_visibility.sql b/schema/postgresql/v12/visibility/versioned/v1.2/advanced_visibility.sql
index b5aeedd3c..45e557a39 100644
--- a/schema/postgresql/v12/visibility/versioned/v1.2/advanced_visibility.sql
+++ b/schema/postgresql/v12/visibility/versioned/v1.2/advanced_visibility.sql
@@ -1,11 +1,3 @@
-CREATE EXTENSION IF NOT EXISTS btree_gin;
-
--- convert_ts converts a timestamp in RFC3339 to UTC timestamp without time zone.
-CREATE FUNCTION convert_ts(s VARCHAR) RETURNS TIMESTAMP AS $$
-BEGIN
-  RETURN s::timestamptz at time zone 'UTC';
-END
-$$ LANGUAGE plpgsql IMMUTABLE RETURNS NULL ON NULL INPUT;
 
 -- Add search attributes related columns
 ALTER TABLE executions_visibility ADD COLUMN search_attributes JSONB NULL;
@@ -15,7 +7,7 @@ ALTER TABLE executions_visibility
   ADD COLUMN TemporalChangeVersion      JSONB         GENERATED ALWAYS AS (search_attributes->'TemporalChangeVersion')                    STORED,
   ADD COLUMN BinaryChecksums            JSONB         GENERATED ALWAYS AS (search_attributes->'BinaryChecksums')                          STORED,
   ADD COLUMN BatcherUser                VARCHAR(255)  GENERATED ALWAYS AS (search_attributes->>'BatcherUser')                             STORED,
-  ADD COLUMN TemporalScheduledStartTime TIMESTAMP     GENERATED ALWAYS AS (convert_ts(search_attributes->>'TemporalScheduledStartTime'))  STORED,
+  ADD COLUMN TemporalScheduledStartTime TIMESTAMP     GENERATED ALWAYS AS (parse_timestamp(search_attributes->>'TemporalScheduledStartTime'))  STORED,
   ADD COLUMN TemporalScheduledById      VARCHAR(255)  GENERATED ALWAYS AS (search_attributes->>'TemporalScheduledById')                   STORED,
   ADD COLUMN TemporalSchedulePaused     BOOLEAN       GENERATED ALWAYS AS ((search_attributes->'TemporalSchedulePaused')::boolean)        STORED,
   ADD COLUMN TemporalNamespaceDivision  VARCHAR(255)  GENERATED ALWAYS AS (search_attributes->>'TemporalNamespaceDivision')               STORED;
@@ -25,9 +17,9 @@ ALTER TABLE executions_visibility
   ADD COLUMN Bool01         BOOLEAN         GENERATED ALWAYS AS ((search_attributes->'Bool01')::boolean)        STORED,
   ADD COLUMN Bool02         BOOLEAN         GENERATED ALWAYS AS ((search_attributes->'Bool02')::boolean)        STORED,
   ADD COLUMN Bool03         BOOLEAN         GENERATED ALWAYS AS ((search_attributes->'Bool03')::boolean)        STORED,
-  ADD COLUMN Datetime01     TIMESTAMP       GENERATED ALWAYS AS (convert_ts(search_attributes->>'Datetime01'))  STORED,
-  ADD COLUMN Datetime02     TIMESTAMP       GENERATED ALWAYS AS (convert_ts(search_attributes->>'Datetime02'))  STORED,
-  ADD COLUMN Datetime03     TIMESTAMP       GENERATED ALWAYS AS (convert_ts(search_attributes->>'Datetime03'))  STORED,
+  ADD COLUMN Datetime01     TIMESTAMP       GENERATED ALWAYS AS (parse_timestamp(search_attributes->>'Datetime01'))  STORED,
+  ADD COLUMN Datetime02     TIMESTAMP       GENERATED ALWAYS AS (parse_timestamp(search_attributes->>'Datetime02'))  STORED,
+  ADD COLUMN Datetime03     TIMESTAMP       GENERATED ALWAYS AS (parse_timestamp(search_attributes->>'Datetime03'))  STORED,
   ADD COLUMN Double01       DECIMAL(20, 5)  GENERATED ALWAYS AS ((search_attributes->'Double01')::decimal)      STORED,
   ADD COLUMN Double02       DECIMAL(20, 5)  GENERATED ALWAYS AS ((search_attributes->'Double02')::decimal)      STORED,
   ADD COLUMN Double03       DECIMAL(20, 5)  GENERATED ALWAYS AS ((search_attributes->'Double03')::decimal)      STORED,
@@ -70,8 +62,8 @@ CREATE INDEX by_history_length  ON executions_visibility (namespace_id, history_
 CREATE INDEX by_task_queue      ON executions_visibility (namespace_id, task_queue,         (COALESCE(close_time, '9999-12-31 23:59:59')) DESC, start_time DESC, run_id);
 
 -- Indexes for the predefined search attributes
-CREATE INDEX by_temporal_change_version       ON executions_visibility USING GIN (namespace_id, TemporalChangeVersion jsonb_path_ops);
-CREATE INDEX by_binary_checksums              ON executions_visibility USING GIN (namespace_id, BinaryChecksums jsonb_path_ops);
+CREATE INDEX by_temporal_change_version       ON executions_visibility USING GIN (namespace_id, TemporalChangeVersion jsonb_ops);
+CREATE INDEX by_binary_checksums              ON executions_visibility USING GIN (namespace_id, BinaryChecksums jsonb_ops);
 CREATE INDEX by_batcher_user                  ON executions_visibility (namespace_id, BatcherUser,                (COALESCE(close_time, '9999-12-31 23:59:59')) DESC, start_time DESC, run_id);
 CREATE INDEX by_temporal_scheduled_start_time ON executions_visibility (namespace_id, TemporalScheduledStartTime, (COALESCE(close_time, '9999-12-31 23:59:59')) DESC, start_time DESC, run_id);
 CREATE INDEX by_temporal_scheduled_by_id      ON executions_visibility (namespace_id, TemporalScheduledById,      (COALESCE(close_time, '9999-12-31 23:59:59')) DESC, start_time DESC, run_id);
@@ -104,6 +96,6 @@ CREATE INDEX by_keyword_10      ON executions_visibility (namespace_id, Keyword1
 CREATE INDEX by_text_01         ON executions_visibility USING GIN (namespace_id, Text01);
 CREATE INDEX by_text_02         ON executions_visibility USING GIN (namespace_id, Text02);
 CREATE INDEX by_text_03         ON executions_visibility USING GIN (namespace_id, Text03);
-CREATE INDEX by_keyword_list_01 ON executions_visibility USING GIN (namespace_id, KeywordList01 jsonb_path_ops);
-CREATE INDEX by_keyword_list_02 ON executions_visibility USING GIN (namespace_id, KeywordList02 jsonb_path_ops);
-CREATE INDEX by_keyword_list_03 ON executions_visibility USING GIN (namespace_id, KeywordList03 jsonb_path_ops);
+CREATE INDEX by_keyword_list_01 ON executions_visibility USING GIN (namespace_id, KeywordList01 jsonb_ops);
+CREATE INDEX by_keyword_list_02 ON executions_visibility USING GIN (namespace_id, KeywordList02 jsonb_ops);
+CREATE INDEX by_keyword_list_03 ON executions_visibility USING GIN (namespace_id, KeywordList03 jsonb_ops);
diff --git a/schema/postgresql/v12/visibility/versioned/v1.3/add_build_ids_search_attribute.sql b/schema/postgresql/v12/visibility/versioned/v1.3/add_build_ids_search_attribute.sql
index 72d1e24e6..f9febd7af 100644
--- a/schema/postgresql/v12/visibility/versioned/v1.3/add_build_ids_search_attribute.sql
+++ b/schema/postgresql/v12/visibility/versioned/v1.3/add_build_ids_search_attribute.sql
@@ -1,2 +1,2 @@
 ALTER TABLE executions_visibility ADD COLUMN BuildIds JSONB GENERATED ALWAYS AS (search_attributes->'BuildIds') STORED;
-CREATE INDEX by_build_ids ON executions_visibility USING GIN (namespace_id, BuildIds jsonb_path_ops);
+CREATE INDEX by_build_ids ON executions_visibility USING GIN (namespace_id, BuildIds jsonb_ops);
```

File: schema/crdb/visibility/schema.sql [hunk 1]
```diff
@@ -0,0 +1,121 @@
+CREATE EXTENSION IF NOT EXISTS btree_gin;
+
+-- convert_ts converts a timestamp in RFC3339 to UTC timestamp without time zone.
+CREATE FUNCTION convert_ts(s VARCHAR) RETURNS TIMESTAMP AS $$
```
Comment (danthegoodman1) [inline schema/crdb/visibility/schema.sql:4]: @Hades32 need to remove this

File: schema/crdb/visibility/schema.sql [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  -- Each search attribute has its own generated column.
+  -- Since PostgreSQL doesn't support virtual columns, all columns are stored.
+  -- PostgreSQL doesn't auto cast to the corresponding column type, so we need to explicitly do it.
+  -- Check the `custom_search_attributes` table for complete set of examples.
+
+  -- Pre-defined search attributes
+  TemporalChangeVersion         JSONB         GENERATED ALWAYS AS (search_attributes->'TemporalChangeVersion')                    STORED,
+  BinaryChecksums               JSONB         GENERATED ALWAYS AS (search_attributes->'BinaryChecksums')                          STORED,
+  BatcherUser                   VARCHAR(255)  GENERATED ALWAYS AS (search_attributes->>'BatcherUser')                             STORED,
+  TemporalScheduledStartTime    TIMESTAMP     GENERATED ALWAYS AS (convert_ts(search_attributes->>'TemporalScheduledStartTime'))  STORED,
```
Comment (danthegoodman1) [inline schema/crdb/visibility/schema.sql:35]: and fix these `conver_ts` functions as well

File: schema/crdb/visibility/schema.sql (hunk 1 above)
Comment (Hades32) [reply schema/crdb/visibility/schema.sql:4]: ah yeah, these files aren't really executed that's why I didn't notice. Will clean it up

**Comment by Hades32** [comment]
we stopped using CRDB. Performance was not great. https://temporalio.slack.com/archives/CTRCR8RBP/p1688653745270049

**Comment by mmack** [comment]
@Hades32 slack is not showing the message anymore. What was your problem with CRDB?

**Pull request closed without merging**